packages/use-i18n/src/index.js

```diff
@@ -0,0 +1,263 @@
+import { formatDistanceToNow, formatDistanceToNowStrict } from 'date-fns'
+import memoizeIntlConstructor from 'intl-format-cache'
+import IntlTranslationFormat from 'intl-messageformat'
+import PropTypes from 'prop-types'
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react'
+import 'intl-pluralrules'
+
+const LOCALE_ITEM_STORAGE = 'locale'
+
+const prefixKeys = prefix => obj =>
+  Object.keys(obj).reduce((acc, key) => {
+    acc[prefix + key] = obj[key]
+
+    return acc
+  }, {})
+
+const I18nContext = createContext()
+
+export const useI18n = () => {
+  const context = useContext(I18nContext)
+  if (context === undefined) {
+    throw new Error('useI18n must be used within a I18nProvider')
+  }
+
+  return context
+}
+
+export const useTranslation = (namespaces = [], load) => {
+  const context = useContext(I18nContext)
+  if (context === undefined) {
+    throw new Error('useTranslation must be used within a I18nProvider')
+  }
+  const { loadTranslations } = context
+
+  const key = namespaces.join(',')
+  useEffect(
+    () =>
+      key.split(',').map(async namespace => loadTranslations(namespace, load)),
+    [loadTranslations, key, load],
+  )
+
+  return context
+}
+
+// https://formatjs.io/docs/intl-messageformat/
+const getTranslationFormat = memoizeIntlConstructor(IntlTranslationFormat)
+const getNumberFormat = memoizeIntlConstructor(Intl.NumberFormat)
+const getDateTimeFormat = memoizeIntlConstructor(Intl.DateTimeFormat)
+const getListFormat = memoizeIntlConstructor(Intl.ListFormat)
+
+const loadDateFnsLocale = async locale =>
+  import(`date-fns/locale/${locale}/index`)
+
+const I18nContextProvider = ({
+  children,
+  defaultLoad,
+  defaultLocale,
+  defaultTranslations,
+  enableDefaultLocale,
+  enableDebugKey,
+  localeItemStorage,
+  supportedLocales,
+}) => {
+  const [currentLocale, setCurrentLocale] = useState(defaultLocale)
+  const [locales, setLocales] = useState(supportedLocales)
+  const [translations, setTranslations] = useState(defaultTranslations)
+  const [dateFnsLocal, setDateFnsLocal] = useState()
+
+  const getLocaleFallback = useCallback(
+    locale => locale.split('-')[0].split('_')[0],
+    [],
+  )
+
+  const getCurrentLocale = useCallback(() => {
+    const languages = navigator.languages || [navigator.language]
+    const browserLocales = [...new Set(languages.map(getLocaleFallback))]
+    const localeStorage = localStorage.getItem(localeItemStorage)
```

Comment:
Do we really need to handle this in this hook ? Shouldn't it be the in the application scope ?

Reply:
i'm agree but maybe in not for this first version. We need to make a lot of change before.